app/src/main/java/com/github/gotify/messages/provider/MessageStateHolder.java

```diff
@@ -9,6 +9,10 @@
     private int lastReceivedMessage = -1;
     private Map<Integer, MessageState> states = new HashMap<>();
 
+    private Message lastRemovedMessage;
+    private int lastRemovedAllPosition;
```

Comment:
Maybe reuse the PositionPair here? Maybe the message could be included in the PositionPair as well, something like MessageWithPosition.

Reply:
I rewrote the logic for the undo state. It was a mess in several ways and should be much cleaner now.